receiver/dotnetdiagnosticsreceiver/dotnet/metadata_parser.go

```diff
@@ -0,0 +1,196 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dotnet
+
+import (
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/dotnetdiagnosticsreceiver/network"
+)
+
+type fieldMsg struct {
+	metadataHeader metadataHeader
+	fields         []field
+}
+
+type metadataHeader struct {
+	metadataID    int32
+	providerName  string
+	eventHeaderID int32
+	eventName     string
+	keyword       uint64
+	version       int32
+	level         int32
+}
+
+// aka payloadFetch
```

Comment:
So, the metadata block has some trickiness to it. When the spec was first created it had no way to represent lists or arrays, I added support for those types for 5.0 (file format v5). In v4 if an event was logged with an array type we would emit the event with empty metadata. Starting in 5 the way it’s handled is we emit the v4 empty metadata and then immediately after but still in the same metadata event we emit a v5 metadata block. 

So what this means is you’ll occasionally see metadata events where the header size plus the payload size do not equal the event size. It’s important when you’re processing events to keep track of the event size in the event header and always seek to it after you’re done, you can’t just assume the event ends after the payload.

Let me know if that doesn’t make sense.

Reply:
Ah, I see. Yeah, I saw that in Microsoft's code but never happened upon a need for it, nor a way to validate any implementation I came up with, so I left it out for now.

Can you help me understand how file format versions relate to .NET versions (if at all -- I'm not a .NET developer so I may be way off 😄) so I can test this?

Reply:
The versions of the file format for 3 and below were called "netperf" files. The only runtime left in support that uses netperf by default is 2.1, and it doesn't have the IPC connection mechanism you are using so you probably don't have to worry about that. Version 4 and above are "nettrace" files and those were introduced for either 3.0 or 3.1, but 3.0 is out of support now. Version 5 of the file format was introduced for .net 5 and that is what contains the metadata stuff mentioned above.

Reply:
If you want to test netperf files on a newer runtime, you can set the environment variable COMPlus_EventPipeNetTraceFormat=0.

Reply:
Here is an eventsource that should trigger the double metadata event in .net 5:

```
    class MyEventSource : EventSource
    {
        public MyEventSource()
            : base(EventSourceSettings.EtwSelfDescribingEventFormat)
        {

        }

        [Event(1, Level = EventLevel.LogAlways)]
        public void BasicEvent(int i)
        {
            WriteEvent(1, i);
        }

        [Event(2, Level = EventLevel.LogAlways)]
        public void StringArrayEvent(string[] args)
        {
            WriteEvent(2, args);
        }

        protected override void OnEventCommand(EventCommandEventArgs command)
        {
            Console.WriteLine($"command={command.Command}");
        }
    }
```

Reply:
Awesome, thank you 👍